### PR TITLE
feat: http payload compression

### LINKED ISSUE #2714: Add Gzip/Brotli compression to REST sink
**What would you like to be added/modified**:

Add Gzip/Brotli, or any other compatible compression algorithm, to REST sink.

**Why is this needed**:

When sending large JSON/XML payload, compressing POST would benefit in terms of latency.

### PR DESCRIPTION
**Pull request by yai-dev**
fix #2714

Below is a details of the changes:

- The `Send` function in `pkg/httpx/http.go` has been refactored. Now we can use the new `HTTPRequestOptions` to customize the request.
- The original headers parameter in the `Send` function has been modified to use `HTTPRequestOptions` with `WithHeadersMap`.
- The original parameters related to Body in the `Send` function have been modified to use `HTTPRequestOptions` with `WithBody`, and parameters related to payload compression: `compressor` and `compressAlgorithm` were provided.
- The `compresssionAlgorithm` parameter has been added to `RawConf` in `internal/io/http/client.go` for setting the compression algorithm, and `compressor` and `decompressor` members have been added to `ClientConf` to store the compressor and decompressor.
- `HTTPLookupSource`, `HTTPPullSource`, and `RestSink` now support data compression and response decompression.
- Update documents for `compressionAlgorithm`.
- A lot of unit tests have been added for all the changes listed above.

BREAKING CHANGE: `pkg/httpx/http.Send` function has changed.

To migrate the code follow the example below:

Before:

`httpx.Send(api.Logger, *http.Client, bodyType, method, url, headers, sendSingle, v)`

After:

`httpx.Send(api.Logger, *http.Client, httpx.WithHeadersMap(headers), httpx.WithBody(v, bodyType, sendSingle, nil, httpx.EmptyCompressorAlgorithm))`

The original function was not friendly enough for scenarios where customization of `http.Request`, so it was modified to use `HTTPRequestOptions` to customization of `http.Request`.

### CODE REVIEW

**Comment by yai-dev** [comment]
I dont know why `data_server_test.go.TestEndpoints` has error in pipeline but worked on my local environment :(

**Comment by ngjaying** [comment]
> I dont know why `data_server_test.go.TestEndpoints` has error in pipeline but worked on my local environment :(

This is a random error. I'll just rerun it.

**Comment by carlclone** [comment]
i think we also need to set content-encoding in request header , so the server can handle the request correctly  . 
reference : [Compressing HTTP requests in Amazon OpenSearch Service](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/gzip.html)

**Comment by yai-dev** [comment]
> i think we also need to set content-encoding in request header , so the server can handle the request correctly . reference : [Compressing HTTP requests in Amazon OpenSearch Service](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/gzip.html)

yep, all that is for compatibility, i will add it.

**Comment by yai-dev** [comment]
@ngjaying i have an idea. should we allow different algorithms for the request and response? for example, the client could use `gzip` for request compression but accept a `zlib` compressed response. the server compresses the response based on the `Accept-Encoding` header, it could make the system more flexible, albeit a bit more complex.

**Comment by ngjaying** [comment]
@yai-dev We are too busy recently to review this PR. Sorry for the delay. Would you like to go through and discuss this PR in tomorrow's weekly meeting?

**Comment by yai-dev** [comment]
> @yai-dev We are too busy recently to review this PR. Sorry for the delay. Would you like to go through and discuss this PR in tomorrow's weekly meeting?

No worries, but i might not be able to attend the weekly meeting tomorrow due to a conflict with work-related metting of my own. If this PR it's not urgent, we could push the PR review to next weekly meeting? Anyway, i really appreciate your invitation to the weekly meeting. 😂

File: .golangci.yaml [hunk 1]
```diff
@@ -3,6 +3,7 @@ run:
   skip-dirs:
     - internal/plugin
     - extensions/functions/tfLite
+    - internal/pkg/httpx/httptestx
```
Comment (ngjaying) [inline .golangci.yaml:6]: why skipping this dir?

File: docs/en_US/guide/sinks/builtin/rest.md [hunk 2]
```diff
@@ -2,21 +2,22 @@
...
+| compressionAlgorithm | true     | Use the specified compression algorithm to compress the payload. Upon receipt of a response, it automatically decompresses the server response using the specified algorithm. The current optional algorithms are: `zlib`, `gzip`, `flate`, `zstd`. **These are case sensitive**. Please note that this option will add the specified algorithm to the request header `Accept-Encoding`. You need to make sure that the server supports the algorithm you set. At the same time, you need to ensure that the server should use the same compression algorithm to compress the response data and set `Content-Encoding` in the response header. For the Flate algorithm, the `Accept-Encoding` request header value will be set to `deflate`. |
```
Comment (ngjaying) [inline docs/en_US/guide/sinks/builtin/rest.md:20]: To comply with other sink like mqtt, suggest to name it compression

File: .golangci.yaml (hunk 1 above)
Comment (yai-dev) [reply .golangci.yaml:6]: cuz the codes under this package uses package with prefix other than `github.com/lf-edge/ekuiper`, the GCI tests cannot pass if these are not ignored.

File: docs/en_US/guide/sinks/builtin/rest.md (hunk 2 above)
Comment (yai-dev) [reply docs/en_US/guide/sinks/builtin/rest.md:20]: okay, working on it

Comment (yai-dev) [reply docs/en_US/guide/sinks/builtin/rest.md]: solved

**Review by ngjaying — APPROVED** [review]

**Pull request merged**